kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/symbol/kotlin/KSTypeImpl.kt

```diff
@@ -62,10 +63,26 @@ class KSTypeImpl private constructor(internal val type: KtType) : KSType {
     }
 
     override val nullability: Nullability by lazy {
-        if (type.nullability == KtTypeNullability.NON_NULLABLE) {
-            Nullability.NOT_NULL
-        } else {
-            Nullability.NULLABLE
+        fun isNullable(type: KtType): Boolean {
+            if (type.nullability == KtTypeNullability.NULLABLE) {
+                return true
+            }
+            if (type is KtFlexibleType && isNullable(type.upperBound)) {
+                return true
+            }
+            if (type is KtDefinitelyNotNullType) {
+                return false
+            }
+            if (type is KtTypeParameterType) {
+                return KSTypeParameterImpl.getCached(type.symbol).bounds.map { it.resolve() }
+                    .contains(ResolverAAImpl.instance.builtIns.anyType.makeNullable())
```

Comment:
1. This goes through KSP's `KSType` creation / lookup
2. This checks if the type parameter has nullable `Any?` as an upper bound, but that misses a lot of nullable type parameters.
3. The whole logic misses `KtIntersectionType`

To mitigate everything all together, how about just using `KtTypeInfoProviderMixIn#canBeNull`?

Re: 2, note that type parameter without any bound is nullable (like just `<T>`); type parameter is nullable if _all_ of its upper bound is nullable (like `<T> ... where T : Int?, T : Long?`, which eventually means T : Number?). The current logic thought the following is nullable `<T> ... where T : Any?, T : Any` which is actually not nullable since upper bounds are intersected. That's the same for intersection type.

Again, `KtTypeInfoProviderMixIn#canBeNull` will take care of everything.

Reply:
The logic is mainly reworked from FE1.0, I guess at least in FE1.0 the check for implicit upper bound is managed somewhere else. But if there is already a utility in AA available, then I am more than happy to use that :)

Reply:
Yeah, AA FE1.0 uses `TypeUtils#isNullableType`: https://github.com/JetBrains/kotlin/blob/master/analysis/analysis-api-fe10/src/org/jetbrains/kotlin/analysis/api/descriptors/components/KtFe10TypeInfoProvider.kt#L41, which will eventually handle implicit upper bound here: https://github.com/JetBrains/kotlin/blob/master/core/descriptors/src/org/jetbrains/kotlin/types/TypeUtils.java#L350